fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ResourceReferenceDAO.java

```diff
@@ -0,0 +1,572 @@
+            // ExternalResourceReferenceRec objects
+            int a = 1;
+            for (String name: systemNames) {
+                ps.setString(a++, name);
+            }
+
+            ResultSet rs = ps.executeQuery();
+            while (rs.next()) {
+                idMap.put(rs.getString(1), rs.getInt(2));
+            }
+        } catch (SQLException x) {
+            logger.log(Level.SEVERE, select.toString(), x);
+            throw translator.translate(x);
+        }
+        
+        // Now update the ids for all the matching systems in our list
+        for (ResourceTokenValueRec xr: systems) {
+            Integer id = idMap.get(xr.getCodeSystemValue());
+            if (id != null) {
+                xr.setCodeSystemValueId(id);
+
+                // Add this value to the (thread-local) cache
+                cache.addCodeSystem(xr.getCodeSystemValue(), id);
+            } else {
+                // Unlikely...but need to handle just in case
+                logger.severe("Record for code_system_name '" + xr.getCodeSystemValue() + "' inserted but not found");
+                throw new IllegalStateException("id deleted from database!");
+            }
+        }
+    }
+    
+    /**
+     * Insert any missing values into the code_systems table
+     * @param paramList
+     * @param systems
+     */
+    public abstract void doCodeSystemsUpsert(String paramList, Collection<String> systemNames);
+    
+    /**
+     * Add reference value records for each unique reference name in the given list
+     * @param values
+     */
+    public void upsertCommonTokenValues(List<ResourceTokenValueRec> values) {
+        
+        // Unique list so we don't try and create the same name more than once.
+        // Ignore any null token-values, because we don't want to (can't) store
+        // them in our common token values table.
+        Set<CommonTokenValue> tokenValues = values.stream().filter(x -> x.getTokenValue() != null).map(xr -> new CommonTokenValue(xr.getCodeSystemValueId(), xr.getTokenValue())).collect(Collectors.toSet());
+        
+        if (tokenValues.isEmpty()) {
+            // nothing to do
+            return;
+        }
+
+        // Build a string of parameter values we use in the query to drive the insert statement.
+        // The database needs to know the type when it parses the query, hence the slightly verbose CAST functions:
+        // VALUES ((CAST(? AS VARCHAR(1234)), CAST(? AS INT)), (...)) AS V(common_token_value, parameter_name_id, code_system_id)
+        StringBuilder inList = new StringBuilder(); // for the select query later
+        StringBuilder paramList = new StringBuilder();
+        for (int i=0; i<tokenValues.size(); i++) {
+            if (paramList.length() > 0) {
+                paramList.append(", ");
+            }
+            paramList.append("(CAST(? AS VARCHAR(" + FhirSchemaConstants.MAX_TOKEN_VALUE_BYTES + "))");
```

Comment:
Does this truncate without warning?

Reply:
How do you mean without warning?

Reply:
If the value is too long, it will generate a SQLException (which may not happen until the value gets inserted into the actual table, but it will happen).